### ISSUE TITLE
Not working with LG Ultrafine 5k

### ISSUE THREAD

**Issue by flybayer** [alin23/Lunar#30]
Hello, I was super excited to find Lunar as I want my LG Ultrafine 5K to auto sync with my macbook pro based on ambient light.

## Problem

In any mode (Manual, Sync, etc), my LG Ultrafine 5K briefly flickers to the brightness set by Lunar but immediately reverts to the brightness as set by macOS.

I'm not sure if there's something on my end? Any suggestions on debugging this? Thanks!

**Comment by alin23** [comment]
Is the monitor connected by direct USB-C connection? I have an LG 4k connected like that and it works perfectly so your monitor should also work.

The Ultrafine has an integrated ambient light sensor, maybe macOS is already using it to auto adjust brightness. Check in Display settings in macOS to make sure any functionality like that is disabled. 

If you have the LG software installed on your Mac, also check to see if it has any auto adjustment settings.

The monitor OSD menu may also have a setting for that, please check that too. 

I don't think there's a problem with Lunar here, it sound like something else is continuously changing the monitor brightness. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by flybayer** [comment]
Thanks! It appears a mac reboot fixed issue I was having.

In the process, I also discovered Luner + F.lux create a similar issue but the display reset is 10-20 seconds instead of <1s.

**Comment by alin23** [comment]
From my understanding, flux changes the gamma tables and color values through the system GPU SDK. Lunar doesn't touch that so they shouldn't interfere in any way. Maybe flux started using DDC too, I don't know as I haven't used it since night shift has been introduced in macOS

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by matej** [comment]
My apologies for hijacking this closed ticket, but I believe there are still issues with the LG Ultrafine 5k. At least with mine.

The "Sync" option doesn't seem to be functioning reliably for my case. It kicks in and syncs the brightens but then soon after it gets overridden back to the initial value. The behavior can differ a bit. Sometimes it gets into a state where it constantly flickers. Sometimes it switches every few seconds. A restart did not affect the behavior for me.

Video: [Lunar5k.mp4.zip](https://github.com/alin23/Lunar/files/3589631/Lunar5k.mp4.zip)

Or perhaps this could be another issue with my particular setup?

I tried to check this with the debugger, but getting some issues with generating the Secrets swift file.

```
Key APIKey lacks a corresponding value
make: *** [Lunar/Generated/Secrets.generated.swift] Error 1
```

**Comment by alin23** [comment]
The LG 5K has an integrated ambient light sensor which is used internally (by macOS or monitor firmware, not sure which) to adjust the brightness. So if you want to use Lunar instead of that, you have to find a way to disable the internal adaptive brightness option. I don't know how to do that because I don't have that monitor. 

But I'm sure that it's not a problem with Lunar. I never write code to create compatibility for specific monitors, I just use the standard DDC protocol. If some monitor/adapter/hub/cable doesn't comply with DDC, then I can't do much to fix that.

Lunar is already working at the lowest possible level that macOS SDK allows, by sending the DDC messages to the GPU driver through IOKit. So if that implementation is correct and works on some monitors, then that's about it. Whatever doesn't work happens most probably because of something out of the software scope, like incomplete cables/hubs without the required pins for DDC, converters that can't pass the DDC information in any way, monitors that have their own logic for brightness/contrast adjustments etc.

Sorry about `make dev` not working. I'll fix that today.

**Comment by alin23** [comment]
@matej this is what the LG UltraFine manual says on page 25:

> • The ambient light sensor operates in accordance with Mac OS. Some Mac OS doesn't
support the display ALS(Ambient Light Sensor).
• You can also adjust the monitor brightness by controlling the menu or hot key of a
host manually

[LG UltraFine 5K manual](https://www.specsserver.com/CACHE/frfmbxeqrhrp.pdf)

**Comment by matej** [comment]
> The LG 5K has an integrated ambient light sensor which is used internally (by macOS or monitor firmware, not sure which) to adjust the brightness. 

No, apparently that's not used. At least I never noticed that the monitor would adjust its brightens, even when manually covering the sensor (on its own or via macOS). And that's basically also confirmed by reports I've been seeing online ([one example](https://discussions.apple.com/thread/250411765)). 

But I understand the rest of your argument completely. I wanted to ask if a behavior like that could have some other known causes, considering the report here says it worked after a reboot and that's not the case for me. 

I'll see if I can find anything out once I get this working with Xcode.

**Comment by matej** [comment]
Oh, I just tried https://github.com/ocjvandijk/Brightness-Sync, from in the discussion I linked before. And that does seem to work. It's a bit choppy and doesn't update the UI in system preferences. But the LG does dim with the MacBook Pro screen.

**Comment by alin23** [comment]
I took a look at BrightnessSync's code and looks like something I could implement in Lunar. The thing is that the code uses a private framework, which isn't accepted by the App Store guidelines and I'd like to get Lunar on the App Store in the near future. 

I'll probably make a non App Store version that will contain these workarounds so that you and other users can keep using Lunar. Thanks for letting me know about this, I'll do my best to come with an update to fix the issue.

**Comment by matej** [comment]
Oh, that's great to hear. Thanks for looking into this!

**Comment by alin23** [comment]
@matej @flybayer I planned to release the new version with the custom code for LG UltraFine monitors but I didn't have time to include that too.

I already started working on it, so it will be released in the near future. Sorry for the delay.

**Comment by matej** [comment]
No worries. Been using Brightness-Sync in the meantime and it works pretty well. So no need to rush because of me.

**Comment by alin23** [comment]
@flybayer @matej can one of you help me test a version of Lunar with support for UltraFine monitors?

You can download it here: https://send.firefox.com/download/d503a7ce1d083d76/#HFsMplUq_ch2JAR-KCb7Yw

**Comment by alin23** [comment]
Go straight to 2.9.3 if you're testing this to avoid a system freeze bug I introduced in 2.9.1

**Comment by matej** [comment]
I gave it a try. The default brightness is off. The LG display noticeably dimmer then the MacBook display. And the transition is a bit choppy. Here's a video to illustrate. Video: [IMG_4216.mp4.zip](https://github.com/alin23/Lunar/files/3909846/IMG_4216.mp4.zip)

Brightness-Sync with the latest updates does a better job syncing up the actual brightens, even without any offset changes.

**Comment by alin23** [comment]
@matej do you have Smooth Transition turned on in Lunar preferences? Also can you write here the values you have for `Brightness offset` and `Contrast offset`?

The defaults may not be appropriate for UltraFine displays and I'll have to adjust them just for this type of monitors.

**Comment by alin23** [comment]
I'll close this issue as UltraFine support has been implemented.

@matej to get the best experience, adjust the Polling Interval to 1 second as UltraFine monitors have a fast response, and play with the min/max contrast and brightness/contrast offsets. 

I have set Smooth Transition as enabled by default for UltraFine displays and adjusted the default values, but if you already started Lunar at least one time before, then the defaults won't change your settings, so you'll have to adjust the offsets and limits by yourself.

But if Brightness-Sync fits your needs well enough, then stick with that as it is more UltraFine specific and it might be a bit faster as it doesn't try to be smart.
